--- nova-2012.1/nova/db/sqlalchemy/models.py.orig   2012-06-03 11:02:20.680836101 +0200
+++ nova-2012.1/nova/db/sqlalchemy/models.py     2012-06-03 11:02:20.590757676 +0200
@@ -344,7 +344,7 @@ def name(self):
     user_id = Column(String(255))
     project_id = Column(String(255))

-    snapshot_id = Column(String(255))
+    snapshot_id = Column(Integer)

     host = Column(String(255))  # , ForeignKey('hosts.id'))
     size = Column(Integer)
