--- a/nova/virt/libvirt/volume.py.orig	2013-04-12 15:38:38.292116416 +0200
+++ b/nova/virt/libvirt/volume.py	2013-04-12 15:38:33.109120902 +0200
@@ -51,6 +51,23 @@
         conf.target_dev = mount_device
         conf.target_bus = "virtio"
         conf.serial = connection_info.get('serial')
+
+        volume_id = connection_info['data']['volume_id']
+
+        # TODO: ec2_volume_id have to be computed from FLAGS.volume_name_template
+        local_volume_ec2id = "/dev/%s/%s" % (FLAGS.volume_group, ec2utils.id_to_ec2_vol_id(volume_id))
+
+        local_volume_uuid = "/dev/%s/%s" % (FLAGS.volume_group, FLAGS.volume_name_template)
+        local_volume_uuid = local_volume_uuid % volume_id
+        is_local_volume_ec2id = islink(local_volume_ec2id)
+        is_local_volume_uuid = islink(local_volume_uuid)
+
+        if is_local_volume_uuid:
+             conf.source_path = local_volume_uuid
+        elif is_local_volume_ec2id:
+             conf.source_path = local_volume_ec2id
+        else:
+             LOG.debug("Attaching device %s as %s" % (conf.source_path, mount_device))
         return conf
 
     def disconnect_volume(self, connection_info, mount_device):
