From 46497fd0254c4c8201b89937262eab2837309aa6 Mon Sep 17 00:00:00 2001
From: Jaroslav Pulchart <jaroslav.pulchart@gooddata.com>
Date: Tue, 18 Sep 2012 16:23:55 +0200
Subject: [PATCH 6/8] Setting promisc on VLAN bridge.

---
 nova/network/linux_net.py | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/nova/network/linux_net.py b/nova/network/linux_net.py
index 1869857..2c592f8 100755
--- a/nova/network/linux_net.py
+++ b/nova/network/linux_net.py
@@ -957,6 +957,9 @@ class LinuxBridgeInterfaceDriver(LinuxNetInterfaceDriver):
         interface = LinuxBridgeInterfaceDriver.ensure_vlan(vlan_num,
                                                bridge_interface, mac_address)
         LinuxBridgeInterfaceDriver.ensure_bridge(bridge, interface, net_attrs)
+        LOG.debug(_('Setting promisc mode on bridge %s'), bridge)
+        _execute('ip', 'link', 'set',
+                     'dev', bridge, 'promisc', 'on', run_as_root=True)
         return interface
 
     @classmethod
-- 
1.7.11.7

