From befe6d31a3d885cf5679d23ba5d8dc06a8718613 Mon Sep 17 00:00:00 2001
From: Jaroslav Pulchart <jaroslav.pulchart@gooddata.com>
Date: Thu, 27 Sep 2012 14:12:31 +0200
Subject: [PATCH 21/21] 1026029-libvirtError_Domain_not_found-temp_fix

---
 nova/virt/libvirt/connection.py | 14 ++++++++++----
 1 file changed, 10 insertions(+), 4 deletions(-)

diff --git a/nova/virt/libvirt/connection.py b/nova/virt/libvirt/connection.py
index 8a14234..e81e661 100644
--- a/nova/virt/libvirt/connection.py
+++ b/nova/virt/libvirt/connection.py
@@ -799,15 +799,21 @@ class LibvirtConnection(driver.ComputeDriver):
             self.volume_driver_method('connect_volume',
                                       connection_info,
                                       mount_device)
-
-        virt_dom = self._conn.lookupByName(instance['name'])
+        try:
+            virt_dom = self._conn.lookupByName(instance['name'])
+        except Exception, e:
+            virt_dom = None
         # NOTE(itoumsn): Use XML delived from the running instance
         # instead of using to_xml(instance, network_info). This is almost
         # the ultimate stupid workaround.
         if not xml:
-            xml = virt_dom.XMLDesc(0)
+            if virt_dom:
+                xml = virt_dom.XMLDesc(0)
+            else:
+                xml = self.to_xml(instance, network_info)
 
-        self._destroy(instance, network_info, cleanup=False)
+        if virt_dom:
+            self._destroy(instance, network_info, cleanup=False)
         self.plug_vifs(instance, network_info)
         self.firewall_driver.setup_basic_filtering(instance, network_info)
         self.firewall_driver.prepare_instance_filter(instance, network_info)
-- 
1.7.11.4

