From 2cb8559c823e436cef8122db2a251e2ba08f0f6a Mon Sep 17 00:00:00 2001
From: Jaroslav Pulchart <jaroslav.pulchart@gooddata.com>
Date: Tue, 9 Oct 2012 14:16:29 +0200
Subject: [PATCH 8/8] Workaround for bug #1061628

---
diff --git a/nova/service.py b/nova/service.py
index 1da10e6..73b39ca 100644
--- a/nova/service.py
+++ b/nova/service.py
@@ -232,7 +232,7 @@ class Service(object):
         if not topic:
             topic = binary.rpartition('nova-')[2]
         if not manager:
-            manager = FLAGS.get('%s_manager' % topic, None)
+            manager = FLAGS.get('%s_manager' % binary.rpartition('nova-')[2], None)
         if report_interval is None:
             report_interval = FLAGS.report_interval
         if periodic_interval is None:
--
1.7.11.7
