*** setup.py.orig	2013-03-18 17:30:59.000000000 +0100
--- setup.py	2013-03-21 14:55:07.023229644 +0100
*************** setuptools.setup(name='nova',
*** 52,57 ****
--- 52,58 ----
                 'bin/nova-dhcpbridge',
                 'bin/nova-manage',
                 'bin/nova-network',
+                'bin/nova-networkvlan',
                 'bin/nova-novncproxy',
                 'bin/nova-objectstore',
                 'bin/nova-rootwrap',

