--- nova/virt/firewall.py.orig	2013-03-18 17:30:59.000000000 +0100
+++ nova/virt/firewall.py	2013-03-21 15:43:45.340530040 +0100
@@ -389,21 +389,15 @@
                     fw_rules += [' '.join(args)]
                 else:
                     if rule['grantee_group']:
-                        # FIXME(jkoelker) This needs to be ported up into
-                        #                 the compute manager which already
-                        #                 has access to a nw_api handle,
-                        #                 and should be the only one making
-                        #                 making rpc calls.
-                        nw_api = network.API()
                         for instance in rule['grantee_group']['instances']:
-                            nw_info = nw_api.get_instance_nw_info(ctxt,
-                                                                  instance)
-
-                            ips = [ip['address']
-                                for ip in nw_info.fixed_ips()
-                                    if ip['version'] == version]
-
-                            LOG.debug('ips: %r', ips, instance=instance)
+                    	    try:
+                        	ips = [ip['address']
+                            	    for ip in db.fixed_ip_get_by_instance(ctxt, instance['id'])]
+                            except:
+                                ips = []
+                            
+                            LOG.debug('ips: %r', ips)
+                            
                             for ip in ips:
                                 subrule = args + ['-s %s' % ip]
                                 fw_rules += [' '.join(subrule)]
