From b03dbc7840036ddfa40034e59e579df90f10a4ef Mon Sep 17 00:00:00 2001
From: Brano Zarnovican <branislav.zarnovican@gooddata.com>
Date: Thu, 1 Nov 2012 11:20:39 +0100
Subject: [PATCH] FIX: set default value for description variable

---
 nova/volume/netapp.py |    1 +
 1 file changed, 1 insertion(+)

diff --git a/nova/volume/netapp.py b/nova/volume/netapp.py
index 1ba559d..88933fe 100644
--- a/nova/volume/netapp.py
+++ b/nova/volume/netapp.py
@@ -288,6 +288,7 @@ class NetAppISCSIDriver(driver.ISCSIDriver):
         project = volume['project_id']
         display_name = volume['display_name']
         display_description = volume['display_description']
+        description = None
         if display_name:
             if display_description:
                 description = display_name + "\n" + display_description
-- 
1.7.10.4

