From a562c128718d737969ab2db1d07e82debac5be2d Mon Sep 17 00:00:00 2001
From: Brano Zarnovican <branislav.zarnovican@gooddata.com>
Date: Thu, 1 Nov 2012 11:43:20 +0100
Subject: [PATCH] FIX: added optional host argument to ensure_export

---
 nova/volume/netapp.py |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/nova/volume/netapp.py b/nova/volume/netapp.py
index 1ba559d..98e6c2b 100644
--- a/nova/volume/netapp.py
+++ b/nova/volume/netapp.py
@@ -434,7 +434,7 @@ class NetAppISCSIDriver(driver.ISCSIDriver):
             raise exception.Error(msg % name)
         return {'provider_location': lun_id}
 
-    def ensure_export(self, context, volume):
+    def ensure_export(self, context, volume, host=None):
         """Driver entry point to get the iSCSI details about an existing volume."""
-- 
1.7.10.4

