From 02aade27222988a9b0e549e7d304b97144be9566 Mon Sep 17 00:00:00 2001
From: Jaroslav Pulchart <jaroslav.pulchart@gooddata.com>
Date: Wed, 7 Nov 2012 14:32:52 +0100
Subject: [PATCH 28/28] Get size of root block device from mapping table.

---
 nova/virt/libvirt/connection.py | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/nova/virt/libvirt/connection.py b/nova/virt/libvirt/connection.py
index e81e661..8550833 100644
--- a/nova/virt/libvirt/connection.py
+++ b/nova/virt/libvirt/connection.py
@@ -1268,6 +1268,18 @@ class LibvirtConnection(driver.ComputeDriver):
 
         root_fname = hashlib.sha1(str(disk_images['image_id'])).hexdigest()
         size = instance['root_gb'] * 1024 * 1024 * 1024
+        # check for root device size in block_device_mapping, if exists
+        try:
+            for bdm in db.block_device_mapping_get_all_by_instance(context,instance['id']):
+                if bdm['device_name'] in ['/dev/vda', '/dev/vda1', '/dev/sda', '/dev/sda1', 'vda1', 'sda1', 'vda', 'sda' ]:
+                    size = bdm['volume_size'] * 1024 * 1024 * 1024
+                    break
+                else:
+                    continue
+        except:
+            LOG.debug(_("Some error in root block device mapping check."))
+
+        LOG.info(_("Root volume size: %s"), size)
 
         inst_type_id = instance['instance_type_id']
         inst_type = instance_types.get_instance_type(inst_type_id)
-- 
1.7.11.7

