From 0356f0d49f66c08f7beffab6a9d12db47a01eb8c Mon Sep 17 00:00:00 2001
From: Jaroslav Pulchart <jaroslav.pulchart@gooddata.com>
Date: Fri, 9 Nov 2012 17:56:17 +0100
Subject: [PATCH 11/11] Ensure that public ips are at the end of the floating
 list.

---
 nova/network/manager.py | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/nova/network/manager.py b/nova/network/manager.py
index 3485e40..53934fe 100644
--- a/nova/network/manager.py
+++ b/nova/network/manager.py
@@ -275,11 +275,19 @@ class FloatingIP(object):
 
         admin_context = context.get_admin_context()
         try:
-            floating_ips = self.db.floating_ip_get_all_by_host(admin_context,
+            unsorted_floating_ips = self.db.floating_ip_get_all_by_host(admin_context,
                                                                self.host)
         except exception.NotFound:
             return
 
+        # Ensure that public ips are at the end of the list to keep correct order of assigement
+        floating_ips = []
+        for floating_ip in unsorted_floating_ips:
+            if floating_ip.get('pool') == "public":
+                 floating_ips.append(floating_ip)
+            else:
+                 floating_ips.insert(0,floating_ip)
+
         for floating_ip in floating_ips:
             fixed_ip_id = floating_ip.get('fixed_ip_id')
             if fixed_ip_id:
-- 
1.7.11.7

