From 496f29082c1626c7a4f33eb4adcdb3307289d905 Mon Sep 17 00:00:00 2001
From: Brano Zarnovican <branislav.zarnovican@gooddata.com>
Date: Thu, 17 Jan 2013 14:34:57 +0100
Subject: [PATCH] BUGFIX: #1099414 - added qtree removal

---
 nova/volume/netapp.py | 24 ++++++++++++++++++++----
 1 file changed, 20 insertions(+), 4 deletions(-)

diff --git a/nova/volume/netapp.py b/nova/volume/netapp.py
index d71586f..a6acbc6 100644
--- a/nova/volume/netapp.py
+++ b/nova/volume/netapp.py
@@ -434,10 +434,11 @@ class NetAppISCSIDriver(driver.ISCSIDriver):
     def _remove_destroy(self, name, project):
         """Remove the LUN from the dataset, also destroying it.
 
-        Remove the LUN from the dataset and destroy the actual LUN on the
-        storage system.
+        Remove the LUN from the dataset and destroy the actual LUN and Qtree
+        on the storage system.
         """
         lun = self._lookup_lun_for_volume(name, project)
+        lun_details = self._get_lun_details(lun.id)
         member = self.client.factory.create('DatasetMemberParameter')
         member.ObjectNameOrId = lun.id
         members = self.client.factory.create('ArrayOfDatasetMemberParameter')
@@ -448,12 +449,27 @@ class NetAppISCSIDriver(driver.ISCSIDriver):
         try:
             server.DatasetRemoveMember(EditLockId=lock_id, Destroy=True,
                                        DatasetMemberParameters=members)
+            res = server.DatasetEditCommit(EditLockId=lock_id,
+                                           AssumeConfirmation=True)
+        except (suds.WebFault, Exception):
+            server.DatasetEditRollback(EditLockId=lock_id)
+            msg = _('Failed to remove and delete dataset LUN member')
+            raise exception.VolumeBackendAPIException(data=msg)
+
+        for info in res.JobIds.JobInfo:
+            self._wait_for_job(info.JobId)
+
+        # Note: it's not possible to delete Qtree & his LUN in one transaction
+        member.ObjectNameOrId = lun_details.QtreeId
+        lock_id = server.DatasetEditBegin(DatasetNameOrId=lun.dataset.id)
+        try:
+            server.DatasetRemoveMember(EditLockId=lock_id, Destroy=True,
+                                       DatasetMemberParameters=members)
             server.DatasetEditCommit(EditLockId=lock_id,
                                      AssumeConfirmation=True)
         except (suds.WebFault, Exception):
             server.DatasetEditRollback(EditLockId=lock_id)
-            msg = _('Failed to remove and delete dataset member')
-            raise exception.VolumeBackendAPIException(data=msg)
+            LOG.warn(_('Volume LUN deleted, Qtree delete failed, ignoring.'))
 
     def create_volume(self, volume):
         """Driver entry point for creating a new volume."""
-- 
1.8.1.4

