From 554a09ec6cc08ba4941add69233f11bbfddbaa91 Mon Sep 17 00:00:00 2001
From: Brano Zarnovican <branislav.zarnovican@gooddata.com>
Date: Thu, 29 Nov 2012 09:43:58 +0100
Subject: [PATCH 7/7] backport for GDC version of driver API

---
 nova/volume/netapp.py | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/nova/volume/netapp.py b/nova/volume/netapp.py
index 25058fe..1ac069c 100644
--- a/nova/volume/netapp.py
+++ b/nova/volume/netapp.py
@@ -614,15 +614,19 @@ class NetAppISCSIDriver(driver.ISCSIDriver):
         lun = self._lookup_lun_for_volume(name, project)
         return {'provider_location': lun.id}
 
-    def ensure_export(self, context, volume):
+    def ensure_export(self, context, volume, host=None):
         """Driver entry point to get the export info for an existing volume."""
+        if hasattr(volume, 'volume_id'):        # if 'volume' is a snapshot
+            return None
         return self._get_export(volume)
 
-    def create_export(self, context, volume):
+    def create_export(self, context, volume, host=None):
         """Driver entry point to get the export info for a new volume."""
+        if hasattr(volume, 'volume_id'):        # if 'volume' is a snapshot
+            return None
         return self._get_export(volume)
 
-    def remove_export(self, context, volume):
+    def remove_export(self, context, volume, host=None):
         """Driver exntry point to remove an export for a volume.
 
         Since exporting is idempotent in this driver, we have nothing
-- 
1.8.1.4

