--- a/nova/volume/iscsi.py.orig	2013-04-11 09:37:45.000000000 +0200
+++ b/nova/volume/iscsi.py	2013-04-11 09:39:01.116259725 +0200
@@ -159,7 +159,7 @@
 
     def remove_iscsi_target(self, tid, lun, vol_id, **kwargs):
         LOG.info(_('Removing volume: %s') % vol_id)
-        vol_uuid_file = 'volume-%s' % vol_id
+        vol_uuid_file = FLAGS.volume_name_template % vol_id
         volume_path = os.path.join(FLAGS.volumes_dir, vol_uuid_file)
         if os.path.isfile(volume_path):
             iqn = '%s%s' % (FLAGS.iscsi_target_prefix,
