From 8839748dec857af64113388579c2e3f8c4690ff9 Mon Sep 17 00:00:00 2001
From: Jaroslav Pulchart <jaroslav.pulchart@gooddata.com>
Date: Fri, 26 Apr 2013 21:52:41 +0200
Subject: [PATCH] FEATURE: graceful shutdown for power off and reboot

---
 nova/virt/libvirt/driver.py | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/nova/virt/libvirt/driver.py b/nova/virt/libvirt/driver.py
index 9ef21b8..a96b8ce 100644
--- a/nova/virt/libvirt/driver.py
+++ b/nova/virt/libvirt/driver.py
@@ -455,6 +455,19 @@ class LibvirtDriver(driver.ComputeDriver):
         # Otherwise, destroy it
         if virt_dom is not None:
             try:
+                virt_dom.shutdown()
+
+                for x in xrange(FLAGS.libvirt_wait_soft_reboot_seconds):
+                    (state, _max_mem, _mem, _cpus, _t) = virt_dom.info()
+                    state = LIBVIRT_POWER_STATE[state]
+                    LOG.debug(_("Instance shutdown status: %s"), state)
+                    if state in [power_state.SHUTDOWN,power_state.CRASHED]:
+                        LOG.info(_("Instance shutdown successfully."),
+                                 instance=instance)
+                        break
+                    else:
+                        greenthread.sleep(1)
+
                 virt_dom.destroy()
             except libvirt.libvirtError as e:
                 is_okay = False
-- 
1.8.1.4

