diff --git a/nova/tests/test_libvirt_config.py b/nova/tests/test_libvirt_config.py
index 34f34df..cc8050c 100644
--- a/nova/tests/test_libvirt_config.py
+++ b/nova/tests/test_libvirt_config.py
@@ -628,6 +628,7 @@ class LibvirtConfigGuestTest(LibvirtConfigBaseTest):
         obj.os_cmdline = "console=xvc0"
         obj.acpi = True
         obj.apic = True
+        obj.pae = True
 
         disk = config.LibvirtConfigGuestDisk()
         disk.source_type = "file"
@@ -653,6 +654,7 @@ class LibvirtConfigGuestTest(LibvirtConfigBaseTest):
               <features>
                 <acpi/>
                 <apic/>
+                <pae/>
               </features>
               <devices>
                 <disk type="file" device="disk">
@@ -673,6 +675,7 @@ class LibvirtConfigGuestTest(LibvirtConfigBaseTest):
         obj.os_boot_dev = "hd"
         obj.acpi = True
         obj.apic = True
+        obj.pae = True
 
         disk = config.LibvirtConfigGuestDisk()
         disk.source_type = "file"
@@ -696,6 +699,7 @@ class LibvirtConfigGuestTest(LibvirtConfigBaseTest):
               <features>
                 <acpi/>
                 <apic/>
+                <pae/>
               </features>
               <devices>
                 <disk type="file" device="disk">
diff --git a/nova/virt/libvirt/config.py b/nova/virt/libvirt/config.py
index 1daf236..4e5341e 100644
--- a/nova/virt/libvirt/config.py
+++ b/nova/virt/libvirt/config.py
@@ -586,6 +586,7 @@ class LibvirtConfigGuest(LibvirtConfigObject):
         self.cpu = None
         self.acpi = False
         self.apic = False
+        self.pae = False
         self.clock = None
         self.os_type = None
         self.os_loader = None
@@ -623,12 +624,14 @@ class LibvirtConfigGuest(LibvirtConfigObject):
         root.append(os)
 
     def _format_features(self, root):
-        if self.acpi or self.apic:
+        if self.acpi or self.apic or self.pae:
             features = etree.Element("features")
             if self.acpi:
                 features.append(etree.Element("acpi"))
             if self.apic:
                 features.append(etree.Element("apic"))
+            if self.pae:
+                features.append(etree.Element("pae"))
             root.append(features)
 
     def _format_devices(self, root):
diff --git a/nova/virt/libvirt/driver.py b/nova/virt/libvirt/driver.py
index 63bd2fe..954eb18 100644
--- a/nova/virt/libvirt/driver.py
+++ b/nova/virt/libvirt/driver.py
@@ -1827,6 +1827,8 @@ class LibvirtDriver(driver.ComputeDriver):
         if FLAGS.libvirt_type != "lxc" and FLAGS.libvirt_type != "uml":
             guest.acpi = True
             guest.apic = True
+            if instance['architecture'] == "i686":
+                guest.pae = True
 
         clk = config.LibvirtConfigGuestClock()
         clk.offset = "utc"
