From bae222b919df07f3022946d31fbc776cc6caabc0 Mon Sep 17 00:00:00 2001
From: Brano Zarnovican <branislav.zarnovican@gooddata.com>
Date: Wed, 22 May 2013 09:02:49 +0200
Subject: [PATCH 2/2] FIX: PCI-640 - trim firewall log for verbose level

---
 nova/virt/firewall.py | 6 ------
 1 file changed, 6 deletions(-)

diff --git a/nova/virt/firewall.py b/nova/virt/firewall.py
index 95a6af8..f001c4e 100644
--- a/nova/virt/firewall.py
+++ b/nova/virt/firewall.py
@@ -370,8 +370,6 @@ class IptablesFirewallDriver(FirewallDriver):
                                                           security_group['id'])
 
             for rule in rules:
-                LOG.debug(_('Adding security group rule: %r'), rule,
-                          instance=instance)
 
                 if not rule.cidr:
                     version = 4
@@ -412,14 +410,10 @@ class IptablesFirewallDriver(FirewallDriver):
                             except:
                                 ips = []
                             
-                            LOG.debug('ips: %r', ips)
-                            
                             for ip in ips:
                                 subrule = args + ['-s %s' % ip]
                                 fw_rules += [' '.join(subrule)]
 
-                LOG.debug('Using fw_rules: %r', fw_rules, instance=instance)
-
         ipv4_rules += ['-j $sg-fallback']
         ipv6_rules += ['-j $sg-fallback']
 
-- 
1.8.1.4

