From 5baeebc1f158ceda414d1e852cd08720db033602 Mon Sep 17 00:00:00 2001
From: Brano Zarnovican <branislav.zarnovican@gooddata.com>
Date: Fri, 16 Aug 2013 11:09:42 +0200
Subject: [PATCH] BUGFIX: PCI-1958 - zero out only first 1G during delete

---
 nova/volume/driver.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/nova/volume/driver.py b/nova/volume/driver.py
index 6b74466..258c00c 100644
--- a/nova/volume/driver.py
+++ b/nova/volume/driver.py
@@ -186,7 +186,7 @@ class VolumeDriver(object):
         # TODO(ja): reclaiming space should be done lazy and low priority
 
         dev_path = self.local_path(volume, ec2id)
-        self._copy_volume('/dev/zero', dev_path, size_in_g)
+        self._copy_volume('/dev/zero', dev_path, 1)
 
         if os.path.exists(dev_path):
             self._try_execute('dmsetup', 'remove', '-f', dev_path,
-- 
1.8.1.4

