From 8ef8bf71ef362cbe0bc8b5dfc3a5cb168d87660b Mon Sep 17 00:00:00 2001
From: Martin Surovcak <martin.surovcak@gooddata.com>
Date: Mon, 16 Sep 2013 10:55:39 +0200
Subject: [PATCH 2/2] PCI-2069: Remove dnsmasq strict-order option

---
 nova/network/linux_net.py | 1 -
 1 file changed, 1 deletion(-)

diff --git a/nova/network/linux_net.py b/nova/network/linux_net.py
index df10a0e..8e0d26d 100644
--- a/nova/network/linux_net.py
+++ b/nova/network/linux_net.py
@@ -864,7 +864,6 @@ def restart_dhcp(context, dev, network_ref):
     cmd = ['FLAGFILE=%s' % FLAGS.dhcpbridge_flagfile,
            'NETWORK_ID=%s' % str(network_ref['id']),
            'dnsmasq',
-           '--strict-order',
            '--bind-interfaces',
            '--conf-file=%s' % FLAGS.dnsmasq_config_file,
            '--domain=%s' % FLAGS.dhcp_domain,
-- 
1.8.3.2

