From 0e49776df55fc2750d142cc3b68e133d9c1e1ed4 Mon Sep 17 00:00:00 2001
From: Martin Surovcak <martin.surovcak@gooddata.com>
Date: Wed, 26 Mar 2014 11:01:00 +0100
Subject: [PATCH] FEATURE: PCI-3639 Allow change root volume size via nova-api

---
 nova/api/openstack/compute/servers.py | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/nova/api/openstack/compute/servers.py b/nova/api/openstack/compute/servers.py
index 6908262..755d8d6 100644
--- a/nova/api/openstack/compute/servers.py
+++ b/nova/api/openstack/compute/servers.py
@@ -799,6 +799,9 @@ class Controller(wsgi.Controller):
         if self.ext_mgr.is_loaded('os-volumes'):
             block_device_mapping = server_dict.get('block_device_mapping', [])
             for bdm in block_device_mapping:
+                if bdm['volume_id'] == '':
+                    LOG.debug('Setting block device mapping volume to None')
+                    bdm['volume_id'] = None
                 if 'delete_on_termination' in bdm:
                     bdm['delete_on_termination'] = utils.bool_from_str(
                         bdm['delete_on_termination'])
-- 
1.8.5.5

