--- nova-2012.1/nova/__init__.py.orig	2012-04-19 15:52:21.962790746 +0200
+++ nova-2012.1/nova/__init__.py	2012-04-19 15:52:56.315089403 +0200
@@ -36,6 +36,29 @@
 
 
 gettext.install('nova', unicode=1)
+
+import sys
+import pkg_resources
+
+# If there is a conflicting non egg module,
+# i.e. an older standard system module installed,
+# then replace it with this requirement
+def replace_dist(requirement):
+    try:
+        return pkg_resources.require(requirement)
+    except pkg_resources.VersionConflict:
+        e = sys.exc_info()[1]
+        dist=e.args[0]
+        req=e.args[1]
+        if dist.key == req.key and not dist.location.endswith('.egg'):
+            del pkg_resources.working_set.by_key[dist.key]
+            # We assume there is no need to adjust sys.path
+            # and the associated pkg_resources.working_set.entries
+            return pkg_resources.require(requirement)
+
+replace_dist("SQLALchemy >= 0.6.3")
+replace_dist("WebOb >= 1.0")
+
 # NOTE(jkoelker) This configures the root logger if it is not already
 #                configured so messages from logging setup can be written
 #                to the console
